### PR TITLE
Issue #2760825 PR #49 code review comments

### PR DESCRIPTION
**Pull request by Kalaiselvan88**
This PR is just FYI on the updates done. I will update on the documentation on KV module as next step

### CODE REVIEW

File: docs/install.md
Comment (fgm) [inline docs/install.md]: Good news: the PHP requirement is now simpler, as the test suite (which has vastly increased last week) now passes on PHP 7.2 too, so can just require PHP7.0x to 7.2.x.

File: docs/install.md
Comment (fgm) [inline docs/install.md]: This is inconsistent with the "Prerequisites" paragraph, for both MongoDB versions, PHP extension versions, and PHP versions. AFAICS, the correct versions for MongoDB are 3.0 to 4.0, for the extension 1.1.7 and upwards including 1.5.0, and PHP 7.0/7.1/7.2. So the best course is probably to just remove that "nutshell" line

File: docs/install.md
Comment (fgm) [inline docs/install.md]: Probably best to add this as a footnote next to the Drush 8.x reference. Not sure whether or not footnotes need an extension to be added to mkdocs. The material theme has them, at least. Or we could just add extra JS/CSS in the configuration file to have them interpreted without adding an extension to mkdocs itself.

File: docs/install.md
Comment (fgm) [inline docs/install.md]: Just like with any HTML page, avoid meaningless link anchors like "here". Use some meaningful text like "the PHP documentation about the extension".

File: docs/install.md
Comment (fgm) [inline docs/install.md]: meaningful link text here too, probably like "the MongoDB deprecation reference page" or something like that.

File: docs/install.md
Comment (fgm) [inline docs/install.md]: typo: missing space after "per"

File: docs/install.md
Comment (fgm) [inline docs/install.md]: already using Composer instead of using Composer already. More generally, Composer being a brand should be uppercased any time it appears in plain text (not in code)

File: docs/install.md
Comment (fgm) [inline docs/install.md:125]: This bug section should probably a sub-paragraph, to make it clear it isn't part of the main Composer section.

File: docs/install.md
Comment (fgm) [inline docs/install.md]: This is not really about Composer requirements, more about installing using Composer.

File: docs/install.md
Comment (fgm) [inline docs/install.md]: backticks around mongodb/mongodb as everywhere else. Same for composer.json.

File: docs/modules/layout.md
Comment (fgm) [inline docs/modules/layout.md]: The mkdocs branch currently contains code directories too, which should probably just be removed to keep the mkdocs branch minimal, with just docs, and not code which might be obsolete and make the user wonder about the differences.

File: docs/modules/layout.md
Comment (fgm) [inline docs/modules/layout.md]: Needs to be clearer: this is the 8.x-2.x branch, the one used by developers until the mkdocs branch is eventually merged into it (or not ?).

File: docs/modules/layout.md
Comment (fgm) [inline docs/modules/layout.md]: That branch also contains a not-up-to-date mkdocs.yml file and docs/ directory. Which should probably be removed too if we keep the docs and code branches separate.

File: docs/modules/mongodb_watchdog.md
Comment (fgm) [inline docs/modules/mongodb_watchdog.md]: collections with an "s": there are several of them, within one database.

File: docs/modules/mongodb_watchdog.md
Comment (fgm) [inline docs/modules/mongodb_watchdog.md]: probably clearer to write "...a separate database, defined by the ..." ?

File: docs/modules/mongodb_watchdog.md
Comment (fgm) [inline docs/modules/mongodb_watchdog.md]: Missing sub heading explaining this is the configuration for the module, and an intro sentence saying this is in config `mongodb_watchdog.settings`

File: docs/install.md
Comment (fgm) [inline docs/install.md:39]: We also probably need to add something about the optional `mod_unique_id` Apache module used for request tracking in Logger.

File: docs/modules/layout.md
Comment (fgm) [inline docs/modules/layout.md]: Wasn't there a way to have that page be the one for the top-level "Modules" navigation item ?

**Review by fgm — CHANGES_REQUESTED** [review]
Beyond the inline comments, I noticed the 2 links to the PHP library on the "bespoke code" page changed to https://docs.mongodb.com/php-library/current 

The intro to the "bespoke code" section should probably highlight that the module provides version a degree of version independence for the API changes in the PHP library (e.g. count method).

The tests page is not up to date: the namespaces have changed, as the page could include a coverage badge like the README does on the code branch. The code should probably be updated from one of the current kernel tests in mongodb_storage or _watchdog. Changes also exist in the core.phpunit.xml file.

Running tests is now entirely in PHPUnit, so any mention of Simpletest can be removed. The best way to rewrite that section is probably to look at how the .travis.yml file does it, and suggest readers to look at that file too.

Also, no need to create a new PR every time: you can just continue pushing code to an existing branch and the PR will update automatically, even if the PR has been merged.

More generally, I wonder whether it is best to keep the mkdocs branch separate, containing just docs, or to merge it to  8.x-2.x at some point, so the docs can be updated along with the code. What is your advice on this ?

File: docs/install.md
Comment (Kalaiselvan88) [reply docs/install.md]: Since 8.x-2.x is the main branch for this suite, it is better to merge mkdocs to 8.x-2.x once we are done with this ticket and current updates. 
   However in any future documentation related issues, we can mention  that the documentation updates to be done on mkdocs branch rather than in the main branch so that at any point it will be easy for getting the commits together on how the documentation has emerged.

File: docs/modules/layout.md
Comment (Kalaiselvan88) [reply docs/modules/layout.md]: Since 8.x-2.x branch is the default and the one which the users will download, I feel mentioning branch here would not be needed from users point of view. Also since we will be merging mkdocs branch to 8.x-2.x branch, am adding branch details as footnote. Let me know how it looks once I commit

File: docs/modules/layout.md
Comment (Kalaiselvan88) [reply docs/modules/layout.md]: Yes. We need to remove code from mkdocs branch. But, shouldn't we do this after we are done with this ticket and documentation updates? Since I am not sure if we delete it now and when merging it with 8.x-2.x branch would GIT delete the code in  8.x-2.x too?

File: docs/modules/layout.md
Comment (fgm) [reply docs/modules/layout.md]: OK, makes sense since we'll be merging the doc on the main 8.2 branch.

File: docs/modules/layout.md
Comment (fgm) [reply docs/modules/layout.md]: Sure, after the merge.

File: docs/modules/layout.md
Comment (Kalaiselvan88) [reply docs/modules/layout.md]: This would be updated when we merge mkdocs with 8.x-2.x

File: docs/modules/layout.md
Comment (Kalaiselvan88) [reply docs/modules/layout.md]: I have moved the layout section above modules since now we also show the layout for documentation too. It makes more sense to have it after Installation menu common to the suite

File: docs/install.md
Comment (Kalaiselvan88) [reply docs/install.md:39]: It is already available in the mongodb_watchdog.md file under mongodb_watchdog.request_tracking variable

Comment (fgm) [reply docs/install.md:39]: Nice changes, and the footnotes extension appears to work well. I just noticed the modules/mongodb/ page also contains references to the old Simpletest namespaces. The explanation about static::$modules doesn't match the code, either, and I'm not sure it is really applicable: look at the watchdog/storage tests for details. The bit about extending settings no longer reflects the current state of the base test class, which provides a getSettingsArray method for that purpose.

A few typos, too: in the layout page, all comments start as uppercase except 3, which should do the same.

I think it should be ready to merge soon when the test pages are done. Regarding tests, be sure to check the phpunit branch, which contains the latest conversion work dropping Simpletest entirely.

Comment (Kalaiselvan88) [reply docs/install.md:39]: I verified in 2988770-watchdog_test_phpunit branch and also in default 8.x-2.x branch but the static::$modules variable is being used in even watchdog and KV modules. So the current line looks fine to me. It would be better if you can provide more clarity on that point. I have updated the settings extension part.

Comment (fgm) [reply docs/install.md:39]: Just looked at the code again, and the code doesn't add to `static::$modules`, but redefines it entirely in every case as you can see, by just redefining the static property every time. Adding to it as the doc suggests would have to be something like:

```php
class FooTest extends WebTestBase {
  // No static redefined here
  // static $modules = [...whatever...]

  public function __construct($name = NULL, array $data = [], $dataName = '') {
    // But extend the existing list here.
    self::$modules[] = 'foo';
    parent::__construct($name, $data, $dataName);
  }
```

But even then, the static property would only be modified upon the test instance creation, and would risk being modified multiple times if the test is instantiated several times, so I'm not sure it would work if PhpUnit starts using static methods before instantiating the test. IMO it is best to do as the existing tests do, and just redefine the whole list of modules in every test instead of relying on this brittle "extension" mechanism in the constructor. 
So it's mostly a question of formulation, saying the test must define the entire list including the list in the parent tests instead of talking about "adding" instead of "redefining", because what the existing tests do is indeed "redefining". What's your take on this ?

Comment (Kalaiselvan88) [reply docs/install.md:39]: Yes. As per current document the code would be like what you have given above. But as you said, there is possibility that if the test is instantiated multiple types then below line would add the module more than 1 time.

```php
self::$modules[] = 'foo';
```
  So there are 2 ways for this. 

1. Follow above method of adding but also have a `__destruct` method that would re-initialise `self::$modules` by removing the current module name or empty it completely. I am not sure if this would work but just an idea.

2. This is the easy and safe method where it is re-defining the `self::$modules` as done in KV and Watchdog tests.

I would also suggest 2nd option. Let me know your opinion then I will modify docs accordingly

Updated rest of document other than this point. Let me know if any further correction required.

Comment (fgm) [reply docs/install.md:39]: Second option for me too, let's roll that.

File: docs/bespoke.md
Comment (fgm) [inline docs/bespoke.md]: "Refer to".

File: docs/modules/mongodb_storage.md
Comment (fgm) [inline docs/modules/mongodb_storage.md:3]: incorrect link, should be [](), not [][]

File: docs/modules/mongodb_storage.md
Comment (fgm) [inline docs/modules/mongodb_storage.md]: incorrect link : doesn't render properly, see previous comment

File: docs/layout.md
Comment (fgm) [inline docs/layout.md]: skip line between title and text

File: docs/install.md
Comment (fgm) [inline docs/install.md]: skip line between title and text

File: docs/install.md
Comment (fgm) [inline docs/install.md]: skip line between title and text

File: docs/modules/mongodb_watchdog.md
Comment (fgm) [inline docs/modules/mongodb_watchdog.md]: skip line between title and text

File: docs/modules/mongodb_watchdog.md
Comment (fgm) [inline docs/modules/mongodb_watchdog.md]: Drupal 8 no longer has "variables": better avoid the term, which makes one think this is unadapted D7 doc. AIUI the term for this is `configuration entry` (or is it `item` , or `value` ?) with `mongodb_watchdog.settings` being a configuration entity.

File: docs/modules/mongodb_watchdog.md
Comment (fgm) [inline docs/modules/mongodb_watchdog.md]: actually, below the ERROR level, these are no longer errors, but warnings, information... I think we use the term "event" in the code and the rest of the page for these, not "errors".

File: docs/tests.md
Comment (fgm) [inline docs/tests.md]: two skipped lines before heading level 1

File: docs/tests.md
Comment (fgm) [inline docs/tests.md]: Now that we have a Functional test in the mongodb_watchdog module (in the yet unmerged phpunit branch) we also need SIMPLETEST_BASE_URL and SIMPLETEST_DB. And BROWSERTEST_OUTPUT_DIRECTORY is recommended to get test output.

Comment (fgm) [reply docs/tests.md]: Note that BROWSERTEST_OUTPUT_DIRECTORY isn't sufficient on its own : the directory must exist and be writable, AND the phpunit.xml configuration needs to add `printerClass="\Drupal\Tests\Listeners\HtmlOutputPrinter"` to the top-level `phpunit` XML element in the `phpunit.xml` file.

File: docs/tests.md
Comment (fgm) [inline docs/tests.md]: Due to the additional variables made necessary by the Functional test, we have a plural again, sorry.

File: docs/tests.md
Comment (fgm) [inline docs/tests.md]: Good place to add the variables mentioned in previous comments.

**Review by fgm — CHANGES_REQUESTED** [review]
You made good progress on this. Most of the remaining changes are typos or relate to the new Functional test requirements.

File: docs/modules/mongodb_storage.md
Comment (Kalaiselvan88) [reply docs/modules/mongodb_storage.md:3]: Actually it renders properly for me in local and gets redirected. Can you provide the change expected here?

image

File: docs/modules/mongodb_storage.md
Comment (Kalaiselvan88) [reply docs/modules/mongodb_storage.md]: refer my comment above

File: docs/layout.md
Comment (Kalaiselvan88) [reply docs/layout.md]: Do you mean a 2 space indentation for Module layout heading? Could not get this...

File: docs/install.md
Comment (Kalaiselvan88) [reply docs/install.md]: Same as above.... could you please explain what is expected here and in below comments? Do you mean to give a 2 space indentation between heading and the content? if so, then we might need to change all files in the documentation as I see it mostly missing

File: docs/modules/mongodb_storage.md
Comment (fgm) [reply docs/modules/mongodb_storage.md:3]: <del>Well, for me it doesn't render in mkdocs. The problem being (AIUI) that you use two sets of square braces, while links are one set of parentheses and one set of square braces. Apparently the literal version of that explanation was eaten by the Github formatter.</del>: works now in a more recent MkDocs build.

File: docs/modules/mongodb_storage.md
Comment (fgm) [reply docs/modules/mongodb_storage.md]: ditto

File: docs/install.md
Comment (fgm) [reply docs/install.md]: Yes, in most of the pages, any time there is a new heading (#, ## or ###) and the previous line is not itself a higher level heading (# for ##, ## for ###), the first heading is preceded by two blank lines, not just one. This doesn't change the output, but makes the source Markdown easier to read.

The spacing between a heading and the next line is 0 blank line if the next line is a heading too, and 1 blank line if the next line is not a heading. This is already the case in most of these docs.

**Comment by fgm** [comment]
Thanks, I answered on the comments, and merged your existing changes so you can rebase your branch and continue: there will be less changes remaining.

2018-08-07 4:42 GMT+02:00 Kalaiselvan Swamy <notifications@github.com>:

**Review by fgm — APPROVED** [review]

**Comment by fgm** [comment]
Thanks for all the work: it is now merged. Better drop your branch to be sure you now have the latest merged version.

**Pull request merged**
